### PR TITLE
MB-262: Migrating LockBox with EncryptedSharedPref

### PR DESCRIPTION
**Pull request by Mandeepbhalothia**

### CODE REVIEW

File: app/src/main/java/cash/z/ecc/android/db/SharedPreferencesManagerImpl.kt [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            prefs = initializeEncryptedSharedPreferencesManager()
+            if (nonEncryptedPreferences.all.isNotEmpty()) { // will be useful for user upgraded OS from L to >= M
+                // migrate non encrypted shared preferences
+                // to encrypted shared preferences and clear them once finished.
+                nonEncryptedPreferences.copyTo(prefs)
+                nonEncryptedPreferences.clear()
+            }
+        } else {
```
Comment (Mandeepbhalothia) [inline app/src/main/java/cash/z/ecc/android/db/SharedPreferencesManagerImpl.kt:29]: @gmale EncryptedSharedPreferences will work from Build.VERSION_CODES.M. Here I am using Normal SharedPrefers just for now for devices less than Build.VERSION_CODES.M.
What is your call or suggestion for this? @nighthawk24 is suggesting **SqlCipher**, even I haven't gone through this.

**Comment by Mandeepbhalothia** [comment]
@gmale @nighthawk24 Safe place is also needed to call this migration stuff. I prefer to call this before we do any operation from Client.

**Comment by nighthawk24** [comment]
@gmale can you please review the status of this PR? Are you planning to finishing the migration away from LockBox to EncryptedSharedPreference? Is this task going to be part of the NU5 support or will it be pursued later?

**Comment by gmale** [comment]
The intention in the [related issue](https://github.com/zcash/zcash-android-wallet/issues/262) was to only touch code inside the Lockbox module. Meaning, all the "Wallet" code would still interact with an abstract concept of a "lockbox" but that lockbox itself would change what it uses for persisting values. That way, any time we change what the wallet uses to persist values, very few code changes (if any) are required outside of the lockbox module.

I should have made this intention more clear. I tried to call it out as the second bullet
>  Incorporate the new library into the lockbox module 

However, I should have added a lot more detail to express the intention. Specifically that nearly all code changes should only occur within the [lockbox module](https://github.com/zcash/zcash-android-wallet/tree/master/lockbox). Ideally, we could also think about migration from the perspective of the module, rather than adding migration to the [main view model](https://github.com/zcash/zcash-android-wallet/pull/267/files#diff-addd71bc62ac199ea43296ed246eadf53d50073352de4e1c7364ae0029be41f3R66-R181). In that case, we might consider how might we migrate from one lockbox type to another without writing much code outside of the lockbox module?

Perhaps:
1. Change [LockBox.kt](https://github.com/zcash/zcash-android-wallet/blob/master/lockbox/src/main/java/cash/z/ecc/android/lockbox/LockBox.kt) into `SecurePreferencesLockBox.kt`
2. Then add `EncryptedSharedPreferencesLockBox.kt` and have it implement `LockBoxPlugin`
3. Change the [AppModule to inject an instance of the new class](https://github.com/zcash/zcash-android-wallet/blob/695bffab9a8e6dd748242a2d0ac74792b2437e9f/app/src/main/java/cash/z/ecc/android/di/module/AppModule.kt#L39). Rather than `LockBox(appContext)`, have it use `EncryptedSharedPreferencesLockBox(appContext)`
4. Then make a few changes to make migrations easier
    i. Update the [LockBoxPlugin interface](https://github.com/zcash/zcash-android-wallet-plugins/blob/master/src/main/java/cash/z/android/plugin/LockBoxPlugin.kt) to add a `getAll` function that allows enumerating over the entries
    ii. implement `getAll` in `SecurePreferencesLockBox` and `EncryptedSharedPreferencesLockBox`
    iii. add a migration step in the wallet that leverages `getAll` to migrate, which might look something like this: 
```
val entries = SecurePreferencesLockBox(appContext).getAll()
if (entries != null) {
    // run the migration which can probably simply be something like:
    //    entries.forEach { newLockBox[it.key] = it.value }
    migrate(entries)  

    // clear out the old prefs so this code doesn't run again
    SecurePreferences.clearAllValues(appContext)
}
```

### Next Steps
Given that the ECC wallet is going to enter maintenance mode pretty soon, it's probably not worthwhile to make these changes because the new NU5 wallet will not make use of it. My suggestion would be to close this PR and leave the existing lockbox logic as is but ensure that the newer ECC and nighthawk wallets use EncryptedSharedPreferences instead.

**Comment by gmale** [comment]
Closing this for now, for simplicity. This can be reopened in the future, if necessary.

**Pull request closed without merging**